build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -374,6 +374,26 @@ message Action {
   // directory, as well as every subdirectory and content blob referred to, MUST
   // be in the
   // [ContentAddressableStorage][build.bazel.remote.execution.v2.ContentAddressableStorage].
+  //
+  // The build action may assume that all files that are part of the
+  // input root are available for reading (i.e., at least 'r--' or 'r-x'
+  // in terms of UNIX permissions). All directories are available for
+  // writing ('rwx'), meaning that files and directories may be created
+  // and deleted by the build action at any place within the input root.
+  //
+  // In addition to that, the entire input root must be exposed through
```

Comment:
I have a couple concerns with this paragraph as written.

 1. It's very linux-centric right now. We've defined the API in a system agnostic way (see e.g. handling of platform separators), and I consider that an important requirement to keep. So the above paragraph specifying "files are readable" and "directories are writable" is good, and I think the examples reasonable to keep since they're clearly examples; this one is defined in terms of mounts and rename() and whatnot that are *nix specific though.

Can this be generalized to more universal properties that hold across OS? If not, is it really a system-specific constraint and thus not actually appropriate for the API spec?

 2. Does the API require specifying this at all? Similar to my other comment, I'm not actually sure it's a limitation that all execution systems must comply with.

This gets into somewhat philosophical territory that is not perfectly settled to me yet, but at a general level it's necessary for some execution systems to allow things others do not, and also, for execution systems to have room to optimize within the constraints of the API. Take RBE's environment controls, for example: https://cloud.google.com/remote-build-execution/docs/changing-docker-config (if this 404's for you you'll need to join the https://groups.google.com/forum/#!forum/rbe-alpha-customers group to get access to our Alpha docs). Users may opt in to various environment features, each of which may also come with a performance pessimization or other tradeoffs. I could see doing the same for actions requiring the input FS to be a single mount point, or where all input files are distinct (and not hardlinks of each other), but *not* adhering to those constraints by default.

What do you think - is this restriction a must-have constraint of the API? Or more practically terms, is *not* adhering to it going to lead to execution systems that are incompatible for "common case" actions? That's a minimal bar I think I'd draw: if "most" actions can run across implementations of the API, that seems good, even if there are corner cases where certain actions need to be changed before they can be ported. But if the majority of actions require porting, clearly we haven't done a good enough job defining a core standard.

Reply:
Ping?